crates/bvs-vault-bank-tokenized/src/contract.rs

```diff
@@ -0,0 +1,650 @@
+        let router = info.sender;
+
+        let vault_balance = UnderlyingToken::query_balance(&deps.as_ref(), &env)?;
+
+        if amount.0 > vault_balance {
+            return Err(VaultError::insufficient("Not enough balance").into());
+        }
+
+        let transfer_msg = UnderlyingToken::bank_send(deps.storage, &router, amount.0)?;
+
+        let event = Event::new("SlashLocked")
+            .add_attribute("sender", router.to_string())
+            .add_attribute("amount", amount.0.to_string())
+            .add_attribute(
+                "token",
+                UnderlyingToken::get_denom(deps.storage)?.to_string(),
+            );
+
+        Ok(Response::new().add_event(event).add_message(transfer_msg))
+    }
+}
+
+#[entry_point]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<cosmwasm_std::Binary> {
+    match msg {
+        QueryMsg::Shares { staker } => to_json_binary(&vault_query::balance_of(deps, staker)?),
+        QueryMsg::Assets { staker } => {
+            let staker = deps.api.addr_validate(&staker)?;
+            to_json_binary(&vault_query::assets(deps, env, staker)?)
+        }
+        QueryMsg::ConvertToAssets { shares } => to_json_binary(
+            &vault_query::convert_to_underlying_token(deps, env, shares)?,
+        ),
+        QueryMsg::ConvertToShares { assets } => {
+            to_json_binary(&vault_query::convert_to_receipt_token(deps, env, assets)?)
+        }
+        QueryMsg::TotalShares {} => {
+            to_json_binary(&vault_query::total_receipt_token_supply(deps, env)?)
+        }
+        QueryMsg::TotalAssets {} => to_json_binary(&vault_query::total_assets(deps, env)?),
+        QueryMsg::QueuedWithdrawal { staker } => {
+            let staker = deps.api.addr_validate(&staker)?;
+            to_json_binary(&vault_query::queued_withdrawal(deps, staker)?)
+        }
+        QueryMsg::VaultInfo {} => to_json_binary(&vault_query::vault_info(deps, env)?),
+        _ => {
+            // cw20 compliant messages are passed to the `cw20-base` contract.
+            // Except for the `Burn` and `BurnFrom` messages.
+            cw20_base::contract::query(deps, env, msg.try_into().unwrap())
+        }
+    }
+}
+
+mod vault_query {
+    use bvs_vault_bank::bank as UnderlyingToken;
+    use bvs_vault_base::msg::VaultInfoResponse;
+    use bvs_vault_base::{
+        offset,
+        shares::{self, QueuedWithdrawalInfo},
+    };
+    use cosmwasm_std::{Addr, Deps, Env, StdResult, Uint128};
+    use cw20_base::contract::query_balance;
+
+    /// Get receipt token balance of the staker
+    /// Since this vault is tokenized, shares are practically the receipt token.
+    /// Such that querying shares is equivalent to querying the receipt token balance of a
+    /// particular staker/address.
+    /// But we will support this query to keep the API consistent with the non-tokenized vault.
+    /// Hopefully that helps with contract consumer/frontend to minimize code changes.
```

Comment:
```suggestion
    /// This helps with the contract consumer/frontend to minimize code changes.
```

Don't need to state `hopefully` since it's a given.